### ISSUE TITLE
Change of default algorithm may cause problems

### ISSUE THREAD

**Issue by desmoteo** [pallets/itsdangerous#111]
I just wanted to share with you  my experience that the change in the default signing algorithm from HS256 to HS512 can break things in case of JSONWebSignatureSerializer that need to be persistent (e.g. stored in a db). 

On our server, previously generated JWTs started causing BadSignature exceptions, resulting in authentication failure.

**Comment by mitsuhiko** [comment]
Ugh sorry. I did not see that this was changed. @davidism can we roll this back?

**Comment by mitsuhiko** [comment]
So my proposal is to add a `fallback_digest_methods` to `HMACAlgorithm` and `Signer`. Then set that to `[sha512]` and change the default back to `sha1` then make a 1.0.1 release.

**Comment by mitsuhiko** [comment]
Damn. This does not work because our key derivation also uses the hash method. :-/

**Comment by mitsuhiko** [comment]
The only fix would involve moving `derive_key` into the hash algorithm as well and derive different keys for all fallbacks.

**Comment by mitsuhiko** [comment]
This is impossible to fix now without making a massive mess of the API. I'm considering wiping the 1.0 release right now and hoping not enough people upgraded yet and then try to figure out how to deal with this. Thoughts? I definitely do not want to go forward with 512 hashes by default.

**Comment by ThiefMaster** [comment]
https://github.com/search?q=%22itsdangerous%3E%3D1.0%22&type=Commits

when nuking the release we should should probably ask those projects to revert to an older version as well

**Comment by mitsuhiko** [comment]
@ThiefMaster yeah. I'm not sure but this is not a great situation right now :( The way I designed (or misdesigned) the key derivation system leaves no good way now to support both hashes that the default was changed without redesigning the entire API.

That said, it does show that generally fallbacks need to be supported anyways so the positive aspect of this is that it forces that conversation now.

**Comment by mitsuhiko** [comment]
I removed the 1.0.0 release. Will figure out the plan tonight.

**Comment by desmoteo** [comment]
Maybe it's bad practice, but wouldn't it be possible to just select the decoding algorithm according to the header?

**Comment by desmoteo** [comment]
@RR2DO2 so even something like this, with a protection against 'none' algorithm, is not advisable i guess https://github.com/desmoteo/itsdangerous/commit/016705775603190c0d6d4bbaacafbcd5faf48965

**Comment by arianmaykon** [comment]
Had some problems with the upgrade too, Flask requires `>=0.24` so we got some breaking changes, we did adjusted/updated to a non breaking change with the latest versions 1.0.0 but just found out this #112.

:/ Fixing again...

**Comment by kthy** [comment]
I realize your team is probably not having the best of weekends, but could you perhaps find two minutes to put up a brief notice on the [Pallets blog](https://palletsprojects.com/blog/) that the package was pulled?
